src/ast/dibuilderbpf.cpp

```diff
@@ -59,14 +86,182 @@ DIType *DIBuilderBPF::getInt64Ty()
   return types_.int64;
 }
 
+DIType *DIBuilderBPF::getInt128Ty()
+{
+  if (!types_.int128)
+    types_.int128 = createBasicType("int128", 128, dwarf::DW_ATE_signed);
+
+  return types_.int128;
+}
+
+DIType *DIBuilderBPF::getIntTy()
+{
+  if (!types_.int_)
+    types_.int_ = createBasicType("int", 32, dwarf::DW_ATE_signed);
+
+  return types_.int_;
+}
+
 DIType *DIBuilderBPF::getInt8PtrTy()
 {
   if (!types_.int8_ptr)
-    types_.int8_ptr = createPointerType(
-        createBasicType("int8", 8, dwarf::DW_ATE_signed), 64);
+    types_.int8_ptr = createPointerType(getInt8Ty(), 64);
 
   return types_.int8_ptr;
 }
 
+// Create anonymous struct with anonymous fields. It's possible that there will
+// be multiple tuples of the same (duplicated) type but BTF deduplication should
+// take care of that.
+DIType *DIBuilderBPF::CreateTupleType(const SizedType &stype)
+{
+  assert(stype.IsTupleTy());
+
+  SmallVector<Metadata *, 8> fields;
+  for (auto &field : stype.GetFields())
+  {
+    fields.push_back(createMemberType(file,
+                                      "",
+                                      file,
+                                      0,
+                                      field.type.GetSize() * 8,
+                                      0,
+                                      field.offset * 8,
+                                      DINode::FlagZero,
+                                      GetType(field.type)));
+  }
+  DICompositeType *result = createStructType(file,
+                                             "",
+                                             file,
+                                             0,
+                                             stype.GetSize() * 8,
+                                             8,
+                                             DINode::FlagZero,
+                                             nullptr,
+                                             getOrCreateArray(fields));
+  return result;
+}
+
+DIType *DIBuilderBPF::GetType(const SizedType &stype)
+{
+  if (stype.IsByteArray() || stype.IsRecordTy())
```

Comment:
nit: would a switch be nicer here?

Reply:
That's probably a good idea, but this is the way we do it everywhere else. Turning these into switches could be useful to get compiler checks but it can introduce some problems so I'd rather do that in a separate PR.